### ISSUE TITLE
New example: Events Functions Patterns

### ISSUE THREAD

**Issue by arthuro555** [GDevelopApp/GDevelop-examples#123]
## Describe the example

This example shows how to use a few popular programming patterns via events functions for helping to develop a game. The patterns that are shown in this example:
- [X] Encapsulating multiple conditions into one (in this example for input management)
- [X] Encapsulating events into a custom behavior (reusing input management as an example)
- [X] Service Locator pattern (in this example for switching between an AI and a User input for player 2)
- [X] Object-oriented programming (3 example scenes for showing using it generally, how to do inheritance and how to do composition)
- [X] State pattern (Making a platformer FSM with an events based behavior)

## Checklist

- [X] My game has a proper name in the game properties. 
- [X] My game package name behind with `com.example.`.
- [X] My game has all events unfolded.
- [X] I've added myself as the author in the game properties.
- [X] I've included a file called "README.md" with a description in proper English, explaining what this example is doing.
- [X] I confirm that this game and all of its resources can be integrated into this GitHub repository, distributed and MIT licensed.
- [X] I've cleaned unused resources.

## Game folder
[FunctionsPatterns.zip](https://github.com/GDevelopApp/GDevelop-examples/files/7271545/FunctionsPatterns.zip)

**Comment by Midhil457** [comment]
Hi, adding to my first review
- Some topics shown in this example have their wiki pages, (FSM, behavior, functions) so you could add a link to the page if anyones needs a step by step guide

I have added my few suggestions in here:
[FunctionsPatterns2.zip](https://github.com/GDevelopApp/GDevelop-examples/files/7317024/FunctionsPatterns2.zip)

**Comment by arthuro555** [comment]
> The example breaks this by communicating directly to one implementation using scene variables.

I agree that this is suboptimal and it frustrated me when I ended up doing that. I did think of adding an action on the services for that but don't quite remember why i decided not to. I'll look into it again.

> An AI will probably need a state at some point so behaviors would be a better fit.

This is more about showing the pattern than it is about showing the optimal way to code an AI/Human second player. Doing it with behaviors would kind of lose its purpose as it would be more of a State pattern (as you said). Originally i was making an example that makes more sense (a generic networking service that can switch between a P2P Provider and an MQTT provider), but I realized midway that it was way too complex and that since all I want to do is explain the patterns in this example, this input switching would be easier to understand.


> About the OOP examples, why not using objects and behaviors?

I could ask the opposite question; Why would I use objects and behaviors for those? I am assuming you mean using an object instance as an instance of a class defined by a behavior attached to the object. In GDevelop, objects are only visual elements, with the exception of the text entry that many said shouldn't have been made an object in hindsight. It just doesn't make sense to me to use a GDevelop object in that context since it is strictly representing data, not any elements to display.

**Comment by Midhil457** [comment]
Is this submission still being worked on? If not, this will be closed :)

**Comment by arthuro555** [comment]
I wasn't, thanks for the ping. I'm on it.

**Comment by arthuro555** [comment]
There, I addressed the reviews comments that I agreed with:
[FunctionsPatterns.zip](https://github.com/GDevelopApp/GDevelop-examples/files/7715392/FunctionsPatterns.zip)

**Comment by Midhil457** [comment]
Sorry for the late review :sweat_smile: 
- In the Scene `OOP: Inheritence` under the `Spawning of new enemies` group, there is an event that is disabled that I think it's needed
- Is the flexbox extension used anywhere? The project's resolution doesn't change, so I don't see a point in having it anyways
- Unfold all the events

I can approve this after these changes are made. If anyone wants to do a review after me, feel free to do so :)

**Comment by arthuro555** [comment]
> Is the flexbox extension used anywhere? The project's resolution doesn't change, so I don't see a point in having it anyways

I have not checked yet what the usage is but fyi the flex box extension is not for adapting to changes in resolution, it is for positioning objects that may be of different sizes in a way that keeps them all in the bounds of another object.
